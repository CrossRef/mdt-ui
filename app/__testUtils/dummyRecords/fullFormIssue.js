import {fromJS} from 'immutable'

export default {
  publicationDoi: "10.5555/dummypublication",
  issueDoi: "10.5555/fullformissue",

  publicationJson: fromJS({
    "status": "ok",
    "message-type": "get",
    "message-version": "1.0.0",
    "message": {
      "doi": "10.5555/dummypublication",
      "owner-prefix": "10.5555",
      "type": "Publication",
      "title": {
        "title": "Test dummy data"
      },
      "state": {},
      "date": "2018-01-31",
      "deposit-timestamp": "20180131104528685",
      "mdt-version": "0",
      "status": "accepted",
      "content": "<Journal xmlns=\"http://www.crossref.org/xschema/1.1\"><journal_metadata><full_title>Test dummy data</full_title><doi_data><doi>10.5555/dummypublication</doi><resource>http://dummyurl.com</resource></doi_data></journal_metadata></Journal>",
      "contains": [
        {
          "doi": "10.5555/fullformissue",
          "owner-prefix": "10.5555",
          "type": "issue",
          "title": {
            "issue": "1",
            "volume": "2",
            "title": "full form issue"
          },
          "state": {
            "archiveLocation": "Internet Archive"
          },
          "date": "2018-02-13",
          "deposit-timestamp": "",
          "mdt-version": "1",
          "status": "draft",
          "content": "",
          "contains": []
        },
        {
          "doi": "10.5555/babafbabfareb",
          "owner-prefix": "10.5555",
          "type": "article",
          "title": {
            "title": "arhaerhae"
          },
          "state": {
            "freetolicense": "yes"
          },
          "date": "2018-02-06",
          "deposit-timestamp": "",
          "mdt-version": "4",
          "status": "draft",
          "content": "",
          "contains": []
        },
        {
          "doi": "",
          "owner-prefix": "10.5555",
          "type": "issue",
          "title": {
            "issue": "5",
            "volume": "",
            "title": ""
          },
          "state": {},
          "date": "2018-01-31",
          "deposit-timestamp": "",
          "mdt-version": "2",
          "status": "draft",
          "content": "",
          "contains": [
            {
              "doi": "10.5555/artundissdummydata",
              "owner-prefix": "10.5555",
              "type": "article",
              "title": {
                "title": "Article Under Issue Dummy Data"
              },
              "state": {
                "archiveLocation": "LOCKSS",
                "freetolicense": "yes"
              },
              "date": "2018-02-05",
              "deposit-timestamp": "20180131104528685",
              "mdt-version": "2",
              "status": "draft",
              "content": "",
              "contains": []
            }
          ]
        }
      ]
    },
    "normalizedRecords": {
      "{\"issue\":\"5\",\"volume\":\"\",\"title\":\"\"}": {
        "doi": "",
        "owner-prefix": "10.5555",
        "type": "issue",
        "title": {
          "issue": "5",
          "volume": "",
          "title": ""
        },
        "state": {},
        "date": "2018-01-31",
        "deposit-timestamp": "",
        "mdt-version": "2",
        "status": "draft",
        "content": "",
        "contains": [
          {
            "doi": "10.5555/artundissdummydata",
            "owner-prefix": "10.5555",
            "type": "article",
            "title": {
              "title": "Article Under Issue Dummy Data"
            },
            "state": {
              "archiveLocation": "LOCKSS",
              "freetolicense": "yes"
            },
            "date": "2018-02-05",
            "deposit-timestamp": "20180131104528685",
            "mdt-version": "2",
            "status": "draft",
            "content": "",
            "contains": []
          }
        ]
      },
      "10.5555/babafbabfareb": {
        "doi": "10.5555/babafbabfareb",
        "owner-prefix": "10.5555",
        "type": "article",
        "title": {
          "title": "arhaerhae"
        },
        "state": {
          "freetolicense": "yes"
        },
        "date": "2018-02-06",
        "deposit-timestamp": "",
        "mdt-version": "4",
        "status": "draft",
        "content": "",
        "contains": []
      },
      "10.5555/fullformissue": {
        "doi": "10.5555/fullformissue",
        "owner-prefix": "10.5555",
        "type": "issue",
        "title": {
          "issue": "1",
          "volume": "2",
          "title": "full form issue"
        },
        "state": {
          "archiveLocation": "Internet Archive"
        },
        "date": "2018-02-13",
        "deposit-timestamp": "",
        "mdt-version": "1",
        "status": "draft",
        "content": "",
        "contains": []
      }
    }
  }),

  issueJson: fromJS({
    "status": "ok",
    "message-type": "get",
    "message-version": "1.0.0",
    "message": {
      "doi": "10.5555/dummypublication",
      "owner-prefix": "10.5555",
      "type": "Publication",
      "title": {"title":"Test dummy data"},
      "state": {},
      "date": "2018-01-31",
      "deposit-timestamp": "20180131104528685",
      "mdt-version": "0",
      "status": "accepted",
      "contains": [
        {
          "doi": "10.5555/fullformissue",
          "owner-prefix": "10.5555",
          "type": "issue",
          "title": {"issue":"1","volume":"2","title":"full form issue"},
          "state": {"archiveLocation":"Internet Archive"},
          "date": "2018-02-13",
          "deposit-timestamp": "",
          "mdt-version": "1",
          "status": "draft",
          "content": "<?xml version=\"1.0\" encoding=\"UTF-8\"?><crossref xmlns=\"http://www.crossref.org/xschema/1.1\"><journal_issue><contributors><person_name sequence=\"first\" contributor_role=\"editor\"><given_name>Bob</given_name><surname>Sagat</surname><suffix>Jr.</suffix><affiliation>Full House</affiliation><ORCID>https://orcid.org/0000-0002-3843-3472</ORCID><alt-name>Jimmy</alt-name></person_name><person_name sequence=\"additional\" contributor_role=\"author\"><given_name>Billy</given_name><surname>Bob</surname><suffix>Sr.</suffix><affiliation>Alabama</affiliation><ORCID>https://orcid.org/0000-0002-3843-3472</ORCID><alt-name>Joe</alt-name></person_name></contributors><titles><title>full form issue</title></titles><publication_date media_type=\"online\"><month>5</month><day>4</day><year>2015</year></publication_date><publication_date media_type=\"print\"><month>3</month><day>5</day><year>2016</year></publication_date><journal_volume><volume>2</volume><doi_data><doi>10.5555/fullissueformvolume</doi><resource>http://fullissueformvolumeurl.com</resource></doi_data></journal_volume><issue>1</issue><special_numbering>124</special_numbering><archive_locations><archive name=\"Internet Archive\"/></archive_locations><doi_data><doi>10.5555/fullformissue</doi><resource>http://fullformissue.com</resource></doi_data></journal_issue></crossref>",
          "contains": [

          ]
        }
      ]
    }
  }),

  issueProp: fromJS({
    "doi": "10.5555/fullformissue",
    "owner-prefix": "10.5555",
    "type": "issue",
    "title": {
      "issue": "1",
      "volume": "2",
      "title": "full form issue"
    },
    "state": {
      "archiveLocation": "Internet Archive"
    },
    "date": "2018-02-13",
    "deposit-timestamp": "",
    "mdt-version": "1",
    "status": "draft",
    "content": "",
    "contains": []
  })
}