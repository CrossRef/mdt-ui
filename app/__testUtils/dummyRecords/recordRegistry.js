import defaultArticle from './defaultArticle'
import fullFormArticle from './fullFormArticle'


export default {
  [defaultArticle.articleDoi]: defaultArticle.articleJson,
  [defaultArticle.publicationDoi]: defaultArticle.publicationJson,

  [fullFormArticle.articleDoi]: fullFormArticle.articleJson,
  [fullFormArticle.publicationDoi]: fullFormArticle.publicationJson

}
