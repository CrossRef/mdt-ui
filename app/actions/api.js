import withQuery from 'with-query'

import authorizedFetch, {regularFetch} from '../utilities/fetch'
const apiBaseUrl = require('../../deployConfig').apiBaseUrl




export function getCRState () {
  return authorizedFetch(`${apiBaseUrl}/state`, {
    method: 'get',
    headers: {Authorization: localStorage.getItem('auth')}
  })
    .then((response)=> {
      return response.text()
    })
}


export function syncState (postingState) {
  authorizedFetch(`${apiBaseUrl}/state`, {
    method: 'post',
    headers: {Authorization: localStorage.getItem('auth')},
    body: JSON.stringify(postingState)
  })
}



export function login (usr, pwd) {
  return authorizedFetch(`${apiBaseUrl}/login`, {
    method: 'post',
    body: JSON.stringify({usr, pwd})
  })
}


export function searchTitle (query) {
  return authorizedFetch(`${apiBaseUrl}/search?q=${query}`, {
    method: 'get',
    headers: {Authorization: localStorage.getItem('auth')}
  })
    .then((response)=> response.json() )
}



export function searchRecords (query, pubTitle, type) {
  return authorizedFetch(`${apiBaseUrl}/search/works?q=${query}&title=${pubTitle}&type=${type.toLowerCase()}`, {
    method: 'get',
    headers: {Authorization: localStorage.getItem('auth')}
  })
    .then((response)=> response.json() )
}



export function getItem (id, forced, alone) {
  let doi='', pubDoi='', title=''  // api looks for empty or missing fields, not "undefined"
  if(typeof id === 'object') {
    doi = id.doi?id.doi:''
    title = id.title?id.title:''
    pubDoi = id.pubDoi?id.pubDoi:''
  } else if(typeof id === 'string') {
    doi = id
  }
// doi from issues with no DOI comes through as "undefined" API wants it blank. 
  const queryParams = `doi=${doi||''}&pubdoi=${pubDoi}&title=${typeof title === 'object' ? JSON.stringify(title) : title}`

  return authorizedFetch(`${apiBaseUrl}/work?${queryParams}${forced ? `&forced=true` : ''}${alone?`&alone=true`:''} `, {
    method: 'get',
    headers: {Authorization: localStorage.getItem('auth')}
  })
    .then(response => {
      if(response.status !== 200) {
        const error = new Error(`${response.status}: ${response.statusText}`)
        error.status = response.status
        throw error
      }
      return response.json()
    })
}



export function getDraftWorks () {
  return authorizedFetch(`${apiBaseUrl}/work?publication=true`, {
    method: 'get',
    headers: {Authorization: localStorage.getItem('auth')}
  })
    .then(response => {
      return response.json()
    })
}



export function submitItem (publication) {
  return authorizedFetch(`${apiBaseUrl}/work`, {
    method:'post',
    headers: {Authorization: localStorage.getItem('auth')},
    body: JSON.stringify(publication)
  })
    .then((response) => {
      if(response.status !== 202) {
        throw `${response.status}: ${response.statusText}`
      }
    })
}




export function deleteItem (id) {
  let doi, pubDoi, title
  if(typeof id === 'object') {
    doi = id.doi
    title = id.title
    pubDoi = id.pubDoi
  } else if(typeof id === 'string') {
    doi = id
  }

  const queryParams = doi ? `doi=${doi}` : `pubdoi=${pubDoi}&title=${typeof title === 'object' ? JSON.stringify(title) : title}`

  return authorizedFetch(`${apiBaseUrl}/work?${queryParams}`, {
    method: 'delete',
    headers: {Authorization: localStorage.getItem('auth')}
  })
}



export function deposit (cartArray) {
  return authorizedFetch(`${apiBaseUrl}/deposit`, {
    method:'post',
    headers: {Authorization: localStorage.getItem('auth')},
    body: JSON.stringify({
      message: cartArray
    })
  })
    .then(result => {
      if(result.status > 202) throw `Server Error ${result.status}: ${result.statusText}`
      return result.json()
    })
}




export function getDepositHistory (params) {
  return authorizedFetch(withQuery(`${apiBaseUrl}/history`, params), {
    method: 'get',
    headers: {Authorization: localStorage.getItem('auth')}
  })
    .then(depositHistory => depositHistory.json())
}




export function getReference (referencesArray) {
  return Promise.all(referencesArray.map( referenceText =>
    new Promise ((resolve, reject) =>
      authorizedFetch(`${apiBaseUrl}/search/references?q=${encodeURIComponent(referenceText)}`, {
        method: 'get',
        headers: {Authorization: localStorage.getItem('auth')}
      })
        .then( result => result.json()).then( result => resolve(result))
        .catch( e => resolve({ message: [{reference: referenceText}]}))
    )
  ))
}



export function getFormattedReference (doi) {
    return authorizedFetch(`${apiBaseUrl}/search/references?q=${encodeURIComponent(doi)}`, {
      method: 'get',
      headers: {Authorization: localStorage.getItem('auth')}
    })
      .then( result => result.json())
}


export function transferTitle (transferTitleState) {

  return authorizedFetch(`${apiBaseUrl}/transferTitle`, {
    method:'post',
    headers: {Authorization: localStorage.getItem('auth')},
    body: JSON.stringify({
      message:transferTitleState
    })
  })
  .then(result => {
    if(result.status > 202) throw `Server Error ${result.status}: ${result.statusText}`
    return result.json()
  })
}

export function getPublisherName (prefix) {
  return authorizedFetch(`${apiBaseUrl}/transferTitle?prefix=${prefix}`, {
    method: 'get',
    headers: {Authorization: localStorage.getItem('auth')}
  })
      .then(result => result.json())
}
export function getPublishers (work) {
  return authorizedFetch(`${apiBaseUrl}/transferTitle`, {
    method: 'get',
    headers: {Authorization: localStorage.getItem('auth')}
  })
      .then(result => result.json())

}
