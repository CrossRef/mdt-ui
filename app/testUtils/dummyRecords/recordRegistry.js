import defaultArticle from './defaultArticle'


export default {
  [defaultArticle.articleDoi]: defaultArticle.articleJson,
  [defaultArticle.publicationDoi]: defaultArticle.publicationJson,





}
