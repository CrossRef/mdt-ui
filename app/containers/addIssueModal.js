import React, { Component } from 'react'
import is from 'prop-types'
import update from 'immutability-helper'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import { controlModal, getPublications } from '../actions/application'
import AddIssueCard from '../components/AddIssueModal/addIssueCard'
import defaultState from '../components/AddIssueModal/issueDefaultState'
import {finishUpdate, doiEntered, escapeString} from '../utilities/helpers'
import getIssueXml from '../components/AddIssueModal/issueXmlGenerator'
import {asyncValidateIssue} from '../utilities/validation'
import parseXMLIssue from '../utilities/parseXMLIssue'
import * as api from '../actions/api'
import { XMLSerializer, DOMParser } from 'xmldom'



const mapStateToProps = (state, props) => ({
  publication: state.publications[props.pubDoi]
})

const mapDispatchToProps = dispatch => bindActionCreators({
  asyncGetPublications: getPublications,
}, dispatch)



export class AddIssueModal extends Component {

  static propTypes = {
    mode: is.string,
    pubDoi: is.string.isRequired,
    asyncGetPublications: is.func.isRequired,
    publication: is.object.isRequired,
    close: is.func.isRequired //close function and reduxControlModal provided by modal container parent
  }

  constructor (props) {
    super();

    this.state = {...defaultState};
    this.state.mode = props.mode || 'add'
    this.state.ownerPrefix = props.pubDoi.split('/')[0]

    this.state.issue.issueDoi = this.state.ownerPrefix + '/'
    this.state.issue.volumeDoi = this.state.ownerPrefix + '/'

    this.state.errorMessages = []
    this.state.focusedInput = ''
    this.state.activeCalendar = ''
  }


  async componentDidMount () {
    if(this.state.mode === 'edit') {
      let id, Publication, Issue, issueDoiDisabled = false, volumeDoiDisabled = false;

      id = {
        doi: this.props.issue.doi,
        title: this.props.issue.title,
        pubDoi: this.props.publication.message.doi
      }
      Publication = await api.getItem(id)
      Issue = Publication.message.contains[0]
      const savedIssueState = Issue.state || {}

      const version = Issue['mdt-version']

      const {issue, optionalIssueInfo, showSection} = parseXMLIssue(Issue.content, this.state.ownerPrefix)

      if(!doiEntered(issue.issueDoi, this.state.ownerPrefix)) {
        issue.issueDoi = this.state.ownerPrefix + '/'
      } else {
        issueDoiDisabled = true
      }

      if(!doiEntered(issue.volumeDoi, this.state.ownerPrefix)) {
        issue.volumeDoi = this.state.ownerPrefix + '/'
      } else {
        volumeDoiDisabled = true
      }

      const checkDuplicateId = false
      const {validatedPayload} = await this.validation(checkDuplicateId, issue, optionalIssueInfo, issueDoiDisabled)

      const setStatePayload = {
        version: version,
        issueDoiDisabled,
        volumeDoiDisabled,
        issue:  issue,
        optionalIssueInfo: optionalIssueInfo,
        showSection: showSection,
        titleId: JSON.stringify({issue: issue.issue, volume: issue.volume, title: issue.issueTitle}),
        depositTimestamp: Issue['deposit-timestamp'],
        ...validatedPayload
      }

      if(savedIssueState.archiveLocation) {
        setStatePayload.issue.archiveLocation = savedIssueState.archiveLocation
      }

      this.setState(setStatePayload, ()=> this.state.validating = false)
    }
  }


  async validation (
    checkDuplicateId = false,
    issueData = this.state.issue,
    optionalIssueInfo = this.state.optionalIssueInfo,
    issueDoiDisabled = this.state.issueDoiDisabled
  ) {

    const { criticalErrors, warnings, contributors, enableVolumeDoi, issueDoiEntered } =
      await asyncValidateIssue({
        issueData,
        optionalIssueInfo,
        ownerPrefix: this.state.ownerPrefix,
        publicationRecords: this.props.publication.normalizedRecords,
        issueDoiDisabled,
        checkDuplicateId
      })

    const validatedPayload = {
      validating: true,
      error: false,
      errors: {...criticalErrors, ...warnings},
      mode: 'edit',
      optionalIssueInfo: (contributors && contributors.length) ? contributors : optionalIssueInfo
    }
    if(enableVolumeDoi) {
      validatedPayload.volumeDoiDisabled = false
    }
    let valid = true;

    for(const key in warnings) {
      if (warnings[key]) {
        validatedPayload.error = true
      }
    }

    for(const key in criticalErrors) {
      if(criticalErrors[key]) {
        validatedPayload.error = true
        valid = false
      }
    }

    validatedPayload.errorMessages = this.errorUtility.onValidate(validatedPayload.errors, validatedPayload.optionalIssueInfo)

    return {valid, validatedPayload, criticalErrors, issueDoiEntered}
  }


  validate = async () => {
    const title = this.state.issue.issueTitle, issue = this.state.issue.issue, volume = this.state.issue.volume
    const newTitleId = JSON.stringify({issue, volume, title})
    const oldTitleId = this.state.titleId
    const titleIdChanged = oldTitleId !== newTitleId

    if(this.state.mode === 'edit') {
      const {validatedPayload} = await this.validation(titleIdChanged)
      this.setState(validatedPayload)
    }
  }


  save = async () => {
    const title = this.state.issue.issueTitle, issue = this.state.issue.issue, volume = this.state.issue.volume
    const newTitleId = JSON.stringify({issue, volume, title})
    const oldTitleId = this.state.titleId
    const titleIdChanged = oldTitleId !== newTitleId
    const issueDoiDisabled = this.state.issueDoiDisabled

    const {valid, validatedPayload, criticalErrors, issueDoiEntered} =
      await this.validation(titleIdChanged)

    validatedPayload.focusedInput = ''

    if (valid) {
      const issueDoi = issueDoiEntered ? this.state.issue.issueDoi : ''
      const { publication, mode } = this.props

      const issueXML = getIssueXml(this.state)

      let version = this.state.version

      if (mode === 'edit') {
        version = String(Number(this.state.version) + 1)
      }

      const newRecord = {
        'title': {issue: escapeString(issue), volume: escapeString(volume), title: escapeString(title)},
        'date': new Date(),
        'deposit-timestamp': this.state.depositTimestamp || null,
        'doi': issueDoi,
        'owner-prefix': this.state.ownerPrefix,
        'type': 'issue',
        'mdt-version': version,
        'status': 'draft',
        'content': new XMLSerializer().serializeToString(issueXML)
      }

      let state = {}
      let hasState = false
      if(this.state.issue.archiveLocation) {
        state.archiveLocation = this.state.issue.archiveLocation
        hasState = true
      }

      if(hasState) {
        newRecord.state = state
      }

      const submissionPayload = {
        message: {
          ...publication.message,
          contains: [newRecord]
        }
      }

      validatedPayload.titleId = newTitleId

      //check if new titleId or doi, if so, do rename
      if(
        (!issueDoiEntered && oldTitleId && titleIdChanged) ||
        (oldTitleId && !issueDoiDisabled && issueDoiEntered)
      ) {
        const issueInfo = mode === 'edit' ?
          publication.normalizedRecords.find( record =>
            JSON.stringify(record.title) === oldTitleId
          )
        : undefined

        //get article content
        let articles = []
        if(issueInfo && issueInfo.contains && issueInfo.contains.length) {
          for (let article of issueInfo.contains) {
            const getArticleContent = await api.getItem(article.doi)
            const articleContent = getArticleContent.message.contains[0].contains[0]
            articleContent['mdt-version'] = String(Number(articleContent['mdt-version']) + 1)
            articles.push(articleContent)
          }
        }

        //delete old issue
        await api.deleteItem({title: oldTitleId, pubDoi: publication.message.doi})

        //save new issue
        await api.submitItem(submissionPayload)

        //save articles to new issue
        for (let articleContent of articles) {
          newRecord.contains = [articleContent]
          await api.submitItem(submissionPayload)
        }

      } else {
        await api.submitItem(submissionPayload)
      }

      this.props.asyncGetPublications(this.props.publication.message.doi)

      if(issueDoiEntered) {
        validatedPayload.issueDoiDisabled = true
      }
      validatedPayload.version = String(Number(this.state.version) + 1)
      const { confirmationPayload, timeOut } = this.confirmSave(criticalErrors)
      validatedPayload.confirmationPayload = confirmationPayload
      validatedPayload.timeOut = timeOut

      this.setState(validatedPayload)

    } else /*if not valid*/ {
      const { confirmationPayload, timeOut } = this.confirmSave(criticalErrors)
      validatedPayload.confirmationPayload = confirmationPayload
      validatedPayload.timeOut = timeOut
      this.setState(validatedPayload)
    }
  }


  confirmSave = (criticalErrors) => {
    clearTimeout(this.state.timeOut)
    const confirmationPayload = {
      status: 'saveSuccess',
      message: 'Save Complete'
    }

    const errorMessageArray = ['Required to save:']

    const criticalErrorMsg = {
      issueVolume: 'Issue or Volume Number.',
      dupTitleIdIssue: 'Issue/Volume already exists.',
      invalidissuedoi: 'DOI must be valid.',
      invalidIssueDoiPrefix: 'DOI must be valid.',
      dupeissuedoi: 'DOI must be valid.',
      volume: 'Volume Number.'
    }

    for (let error in criticalErrors) {
      if(criticalErrors[error] === true) {
        confirmationPayload.status = 'saveFailed'
        errorMessageArray.push(criticalErrorMsg[error])
      }
    }

    if(confirmationPayload.status === 'saveFailed') {
      confirmationPayload.message = errorMessageArray.join(' ')
    }

    const timeOut = setTimeout(()=>{
      this.setState({confirmationPayload: {status: '', message: ''}})
    }, 7000)

    return {confirmationPayload, timeOut}
  }


  errorUtility = {

    errorIndicators: [],

    activeIndicator: -1,

    openingSubItem: false,
    subItemIndex: "0",

    saveRef: (activeErrors, indicatorErrors, node, subItem, subItemIndex, openSubItem) => {
      if(node){
        this.errorUtility.errorIndicators.push({
          activeErrors,
          indicatorErrors,
          node,
          subItem,
          subItemIndex,
          openSubItem
        })

        if(node.id === 'errorBubble') {
          this.errorUtility.activeIndicator = this.errorUtility.errorIndicators.length - 1
        }
      }
    },

    setErrorMessages: (setErrors, allErrors = this.state.errors) => {
      const filteredErrorMessage = setErrors.filter((error)=>{
        return allErrors[error]
      })
      this.setState({errorMessages: filteredErrorMessage})
    },

    onValidate: (newValidationErrors, optionalIssueInfo) => {
      if(!this.state.errorMessages.length) {
        return []
      }

      const {errorIndicators, activeIndicator} = this.errorUtility
      const activeIndicatorObj = errorIndicators[activeIndicator]
      const trackedIndicatorErrors = activeIndicatorObj ? activeIndicatorObj.indicatorErrors : []
      let newErrorMessages
      const {subItem, subItemIndex} = activeIndicatorObj || {}

      if(subItem) {

        let allErrors = optionalIssueInfo[subItemIndex].errors

        newErrorMessages = this.state.errorMessages.filter((error) => {
          return allErrors[error]
        })

        if(!newErrorMessages.length) {
          newErrorMessages = trackedIndicatorErrors.filter((error) => {
            return allErrors[error]
          })
        }

      } else {
        newErrorMessages = this.state.errorMessages.filter((error) => {
          return newValidationErrors[error]
        })

        if(!newErrorMessages.length) {
          newErrorMessages = trackedIndicatorErrors.filter((error)=>{
            return newValidationErrors[error]
          })
        }

        this.errorUtility.subItemIndex = "0"
      }

      if(!newErrorMessages.length) {
        const indicatorBelow = errorIndicators[activeIndicator + 1]
        const indicatorAbove = errorIndicators[activeIndicator - 1]

        if(indicatorBelow) {
          newErrorMessages = indicatorBelow.activeErrors
          this.errorUtility.subItemIndex = indicatorBelow.subItemIndex || "0"
        } else if(indicatorAbove) {
          newErrorMessages = indicatorAbove.activeErrors
          this.errorUtility.subItemIndex = indicatorAbove.subItemIndex || "0"
        }
      }

      return newErrorMessages
    }
  }


  tooltipUtility = {

    getFocusedInput: () => this.state.focusedInput,

    tooltipMounted: false,

    assignRefreshTask: (func) => {
      this.tooltipUtility.tooltipMounted = true
      this.tooltipUtility.refreshTask = func
    },

    refresh: (param) => {
      if(
        this.tooltipUtility.tooltipMounted &&
        typeof this.tooltipUtility.refreshTask === 'function'
      ) {
        return finishUpdate().then(()=>this.tooltipUtility.refreshTask(param))
      }
    },

    assignFocus: (inputId, tooltip) => {
      this.setState({focusedInput: inputId})
      return this.tooltipUtility.refresh(tooltip)
    }
  }


  handler = (e) => {
    const name = e.target.name

    if (name === 'issueUrl') {
      (e.target.value.trim().length > 0) ? this.setState({showIssueDoiReq:true}) : this.setState({showIssueDoiReq:false})
    }

    this.setState({
      issue: update(this.state.issue, {[name]: {$set: e.target.value ? e.target.value : '' }})
    })
  }


  helperSwitch = () => {
      this.setState({showHelper: !this.state.showHelper})
  }


  boundSetState = (...args) => { this.setState(...args) }


  addSubItem = () => {
    this.setState({
      optionalIssueInfo: update(this.state.optionalIssueInfo, {$push: defaultState.optionalIssueInfo})
    })
  }


  removeSubItem = (index) => {
    this.setState({ // this situation, state did NOT update immediately to see change, must pass in a call back
      optionalIssueInfo: update(this.state.optionalIssueInfo, {$splice: [[index, 1]] })
    })
  }


  componentDidUpdate (prevProps, prevState) {
    //Select first error if first validation
    if(
      this.state.validating &&
      this.state.error && !prevState.error
    ) {
      try {
        this.errorUtility.setErrorMessages(this.errorUtility.errorIndicators[0].activeErrors)
      } catch (e) {}
    }

    this.tooltipUtility.refresh()

    this.state.validating = false
  }


  componentWillUnmount () {
    clearTimeout(this.state.timeOut)
  }


  calendarHandler = (name, dateObj) => {
    if(dateObj) {
      const stateName = this.state.activeCalendar.split('-')[0]
      const datePayload = {
        [`${stateName}Year`]: dateObj.year,
        [`${stateName}Month`]: dateObj.month,
        [`${stateName}Day`]: dateObj.day
      }

      this.setState({
        activeCalendar: name,
        issue: {
          ...this.state.issue,
          ...datePayload
        }
      }, () => this.validate())

    } else {
      this.setState({
        activeCalendar: name
      })
    }
  }


  render () {
    this.errorUtility.errorIndicators = []  //Saving refs of any errorIndicators rendered so need to clear it before each render
    this.errorUtility.openingSubItem = false
    return (
      <AddIssueCard
        save={this.save}
        handler={this.handler}
        addSubItem={this.addSubItem}
        removeSubItem={this.removeSubItem}
        closeModal={this.props.close}
        helperSwitch={this.helperSwitch}
        errorUtility={this.errorUtility}
        validate={this.validate}
        boundSetState={this.boundSetState}
        tooltipUtility={this.tooltipUtility}
        calendarHandler={this.calendarHandler}
        {...this.state}
      />
    )

  }
}


export default connect(mapStateToProps, mapDispatchToProps)(AddIssueModal)

