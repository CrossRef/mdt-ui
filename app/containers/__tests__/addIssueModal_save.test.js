import React from 'react';
import Enzyme, { shallow, mount, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });
import toJson from 'enzyme-to-json';
import sinon from 'sinon'

import * as api from '../../actions/api'
import {AddIssueModal} from '../addIssueModal'
import fullFormIssue from '../../__testUtils/dummyRecords/fullFormIssue'
import recordRegistry from '../../__testUtils/dummyRecords/recordRegistry'





jest.mock("../../actions/api")



const sandbox = sinon.createSandbox()
sandbox.stub(api, 'getItem')
sandbox.stub(api, 'submitItem')



const commonProps = {
  asyncGetPublications: sandbox.stub(),
  close: sandbox.stub()
}


function testSave (testName, {pubDoi, publicationJson, issueProp, mode}) {
  let component
  let instance
  let save

  const issueProps = {
    ...commonProps,
    publication: publicationJson,
    pubDoi: pubDoi,
    issue: issueProp,
    mode: mode
  }

  beforeEach(() => {
    sandbox.resetHistory()
    api.getItem.callsFake( id => Promise.resolve(recordRegistry[id.doi || JSON.stringify(id.title)].toJS()))

    component = mount(
      <AddIssueModal {...issueProps}/>
    )

    instance = component.instance()
  })

  test(testName, async () => {

    api.getItem.callsFake( id => {
      throw "Mock doi dupe check"
    })

    await instance.save()

    const savedJson = api.submitItem.args[0][0]
    delete savedJson.message.date
    delete savedJson.message.contains[0].date
    try {
      delete savedJson.message.contains[0].contains[0].date
    } catch (e){}

    expect(savedJson).toMatchSnapshot()
  })
}



describe('save', ()=>testSave('edit fullFormIssue', {
  pubDoi: fullFormIssue.publicationDoi,
  publicationJson: fullFormIssue.publicationJson.toJS(),
  issueProp: fullFormIssue.issueProp.toJS(),
  mode: 'edit'
}))