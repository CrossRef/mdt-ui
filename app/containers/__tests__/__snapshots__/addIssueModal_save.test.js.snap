// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`save edit fullFormIssue 1`] = `
Object {
  "message": Object {
    "contains": Array [
      Object {
        "content": "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><crossref xmlns=\\"http://www.crossref.org/xschema/1.1\\"><journal_issue><contributors><person_name sequence=\\"first\\" contributor_role=\\"editor\\"><given_name>Bob</given_name><surname>Sagat</surname><suffix>Jr.</suffix><affiliation>Full House</affiliation><ORCID>https://orcid.org/0000-0002-3843-3472</ORCID><alt-name>Jimmy</alt-name></person_name><person_name sequence=\\"additional\\" contributor_role=\\"author\\"><given_name>Billy</given_name><surname>Bob</surname><suffix>Sr.</suffix><affiliation>Alabama</affiliation><ORCID>https://orcid.org/0000-0002-3843-3472</ORCID><alt-name>Joe</alt-name></person_name></contributors><titles><title>full form issue</title></titles><publication_date media_type=\\"online\\"><month>5</month><day>4</day><year>2015</year></publication_date><publication_date media_type=\\"print\\"><month>3</month><day>5</day><year>2016</year></publication_date><journal_volume><volume>2</volume><doi_data><doi>10.5555/fullissueformvolume</doi><resource>http://fullissueformvolumeurl.com</resource></doi_data></journal_volume><issue>1</issue><special_numbering>124</special_numbering><archive_locations><archive name=\\"Internet Archive\\"/></archive_locations><doi_data><doi>10.5555/fullformissue</doi><resource>http://fullformissue.com</resource></doi_data></journal_issue></crossref>",
        "deposit-timestamp": null,
        "doi": "10.5555/fullformissue",
        "mdt-version": "2",
        "owner-prefix": "10.5555",
        "state": Object {
          "archiveLocation": "Internet Archive",
        },
        "status": "draft",
        "title": Object {
          "issue": "1",
          "title": "full form issue",
          "volume": "2",
        },
        "type": "issue",
      },
    ],
    "content": "<Journal xmlns=\\"http://www.crossref.org/xschema/1.1\\"><journal_metadata><full_title>Test dummy data</full_title><doi_data><doi>10.5555/dummypublication</doi><resource>http://dummyurl.com</resource></doi_data></journal_metadata></Journal>",
    "deposit-timestamp": "20180131104528685",
    "doi": "10.5555/dummypublication",
    "mdt-version": "0",
    "owner-prefix": "10.5555",
    "state": Object {},
    "status": "accepted",
    "title": Object {
      "title": "Test dummy data",
    },
    "type": "Publication",
  },
}
`;
